### ISSUE TITLE
predictor_exporter: Parameter to MergeFrom() must be instance of same class

### ISSUE THREAD

**Issue by gjpicker** [facebookarchive/caffe2#1573]
1. python command 
`
  LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/username/workspace/caffe2_conda_3/lib/  PYTHONPATH=/home/username/workspace/caffe2_conda_3/  /home/username/workspace/caffe2_install/bin/python   resnet50_trainer.py --train_data dataset/cifar10_train_lmdb --test_data dataset/cifar10_test_lmdb --gpus 0,1,2,3 --batch_size 128 --num_labels 10 --epoch_size 50000 --num_epochs 10 --num_shards 2 --shard_id 1 --run_id 50080 --redis_host 10.38.9.197
`

2. stack error 
```
INFO:resnet50_trainer:Training loss: 2.13840866089, accuracy: 0.21875
INFO:resnet50_trainer:Finished iteration 191/195 of epoch 0 (178.87 images/sec)
INFO:resnet50_trainer:Training loss: 2.13097214699, accuracy: 0.125
INFO:resnet50_trainer:Finished iteration 192/195 of epoch 0 (178.98 images/sec)
INFO:resnet50_trainer:Training loss: 2.06589603424, accuracy: 0.15625
INFO:resnet50_trainer:Finished iteration 193/195 of epoch 0 (178.42 images/sec)
INFO:resnet50_trainer:Training loss: 2.06455469131, accuracy: 0.1875
INFO:resnet50_trainer:Finished iteration 194/195 of epoch 0 (175.69 images/sec)
INFO:resnet50_trainer:Training loss: 2.13960528374, accuracy: 0.3125
INFO:resnet50_trainer:Finished iteration 195/195 of epoch 0 (179.28 images/sec)
INFO:resnet50_trainer:Training loss: 2.00170516968, accuracy: 0.28125
Traceback (most recent call last):
  File "resnet50_trainer.py", line 623, in <module>
    main()
  File "resnet50_trainer.py", line 619, in main
    Train(args)
  File "resnet50_trainer.py", line 541, in Train
    SaveModel(args, train_model, epoch)
  File "resnet50_trainer.py", line 137, in SaveModel
    predictor_export_meta=predictor_export_meta,
  File "/home/username/workspace/caffe2_conda_3/caffe2/python/predictor/predictor_exporter.py", line 191, in save_to_db
    meta_net_def = get_meta_net_def(predictor_export_meta)
  File "/home/username/workspace/caffe2_conda_3/caffe2/python/predictor/predictor_exporter.py", line 169, in get_meta_net_def
    utils.create_predict_init_net(ws, predictor_export_meta))
  File "/home/username/workspace/caffe2_conda_3/caffe2/python/predictor/predictor_py_utils.py", line 137, in AddNet
    meta_net_def.nets.add(key=net_name, value=net_def)
TypeError: Parameter to MergeFrom() must be instance of same class: expected caffe2.NetDef got caffe2.NetDef.
```

3. to avoid bug ,if I delete  line 134~line 138  ,it run success .
commit/7facfa4a841cb78a40267701b813d0286a267cbd
resnet50_trainer.py

>>>pred_exp.save_to_db(
>>>    db_type="minidb",
>>>    db_destination=model_path,
>>>    predictor_export_meta=predictor_export_meta,
>>>  )

```
INFO:resnet50_trainer:Finished iteration 193/195 of epoch 0 (177.80 images/sec)
INFO:resnet50_trainer:Training loss: 2.31114125252, accuracy: 0.125
INFO:resnet50_trainer:Finished iteration 194/195 of epoch 0 (178.93 images/sec)
INFO:resnet50_trainer:Training loss: 2.30473995209, accuracy: 0.0625
INFO:resnet50_trainer:Finished iteration 195/195 of epoch 0 (178.54 images/sec)
INFO:resnet50_trainer:Training loss: 2.28475117683, accuracy: 0.0625


INFO:resnet50_trainer:Starting epoch 1/10
INFO:resnet50_trainer:Finished iteration 1/195 of epoch 1 (179.22 images/sec)
INFO:resnet50_trainer:Training loss: 2.3367831707, accuracy: 0.0
INFO:resnet50_trainer:Finished iteration 2/195 of epoch 1 (179.61 images/sec)
INFO:resnet50_trainer:Training loss: 2.29775261879, accuracy: 0.125
INFO:resnet50_trainer:Finished iteration 3/195 of epoch 1 (179.27 images/sec)
INFO:resnet50_trainer:Training loss: 2.31890296936, accuracy: 0.09375
INFO:resnet50_trainer:Finished iteration 4/195 of epoch 1 (179.22 images/sec)
```

ps.   cmake  config  same as  #1553

**Comment by pietern** [comment]
@salexspb Does this ring a bell for you?

**Comment by salexspb** [comment]
hm, not really. Looks like some protobuf issue to me, error message is confusing. 
Do we have a container to repro on?

**Comment by gjpicker** [comment]
@salexspb 
yes, if I upgrade protobuf to same version ,bug fixed ! 

host_1 
>> pip list 
`
protobuf(3.4.0) ---->   protobuf-3.5.0.post1-cp27-cp27mu-manylinux1_x86_64.whl
`

host_2
>> pip list 
`
protobuf (3.5.0.post1)
`

**Comment by sclarkson** [comment]
I see this error with a fresh conda install

`conda list | grep caffe2`

gives

`caffe2                    0.8.dev          py36h650a854_0    caffe2`

```
pytest -x anaconda3/lib/python3.6/site-packages/caffe2/python/predictor/predictor_exporter_test.py
```

gives

```
E       TypeError: Parameter to MergeFrom() must be instance of same class: expected caffe2.NetDef got caffe2.NetDef.
```

Full output:

```
================================================= test session starts =================================================
platform linux -- Python 3.6.4, pytest-3.3.2, py-1.5.2, pluggy-0.6.0
rootdir: /home/sclarkson, inifile:
collected 5 items                                                                                                     

anaconda3/lib/python3.6/site-packages/caffe2/python/predictor/predictor_exporter_test.py .F

====================================================== FAILURES =======================================================
____________________________________ PredictorExporterTest.test_load_device_scope _____________________________________

self = <caffe2.python.predictor.predictor_exporter_test.PredictorExporterTest testMethod=test_load_device_scope>

    def test_load_device_scope(self):
        for param, value in self.params.items():
            workspace.FeedBlob(param, value)
    
        pem = pe.PredictorExportMeta(
            predict_net=self.predictor_export_meta.predict_net,
            parameters=self.predictor_export_meta.parameters,
            inputs=self.predictor_export_meta.inputs,
            outputs=self.predictor_export_meta.outputs,
            shapes=self.predictor_export_meta.shapes,
            net_type='dag',
        )
    
        db_type = 'minidb'
        db_file = tempfile.NamedTemporaryFile(
            delete=False, suffix=".{}".format(db_type))
        pe.save_to_db(
            db_type=db_type,
            db_destination=db_file.name,
>           predictor_export_meta=pem)

anaconda3/lib/python3.6/site-packages/caffe2/python/predictor/predictor_exporter_test.py:178: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
anaconda3/lib/python3.6/site-packages/caffe2/python/predictor/predictor_exporter.py:201: in save_to_db
    meta_net_def = get_meta_net_def(predictor_export_meta)
anaconda3/lib/python3.6/site-packages/caffe2/python/predictor/predictor_exporter.py:179: in get_meta_net_def
    utils.create_predict_init_net(ws, predictor_export_meta))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

meta_net_def = nets {
  key: "PREDICT_INIT_NET_TYPE"
}
, net_name = 'PREDICT_INIT_NET_TYPE'
net_def = name: "predict-init_1"
op {
  output: "data"
  name: ""
  type: "ConstantFill"
  arg {
    name: "shape"
    ints: 1
 ...AddToSelf\n    op = CreateOperator(op_type, inputs, outputs, **kwargs)\n"
}
external_input: "data"
external_input: "y"


    def AddNet(meta_net_def, net_name, net_def):
>       meta_net_def.nets.add(key=net_name, value=net_def)
E       TypeError: Parameter to MergeFrom() must be instance of same class: expected caffe2.NetDef got caffe2.NetDef.

anaconda3/lib/python3.6/site-packages/caffe2/python/predictor/predictor_py_utils.py:142: TypeError
------------------------------------------------ Captured stderr call -------------------------------------------------
WARNING:root:[====DEPRECATE WARNING====]: you are creating an object from CNNModelHelper class which will be deprecated soon. Please use ModelHelper object with brew module. For more information, please refer to caffe2.ai and python/brew.py, python/brew_test.py for more information.
cnn.py                      40 WARNING  [====DEPRECATE WARNING====]: you are creating an object from CNNModelHelper class which will be deprecated soon. Please use ModelHelper object with brew module. For more information, please refer to caffe2.ai and python/brew.py, python/brew_test.py for more information.

-------------------------------------------------- Captured log call --------------------------------------------------
cnn.py                      40 WARNING  [====DEPRECATE WARNING====]: you are creating an object from CNNModelHelper class which will be deprecated soon. Please use ModelHelper object with brew module. For more information, please refer to caffe2.ai and python/brew.py, python/brew_test.py for more information.
========================================= 1 failed, 1 passed in 0.29 seconds ==========================================
```

**Comment by pjh5** [comment]
@sclarkson You can work around this by copy-pasting caffe2/caffe2/proto/metanet.proto into the bottom of caffe2/caffe2/proto/caffe2.proto (copy only the message objects, so don't copy the first 5 lines of the file) and then renaming all references of "metanet_pb2" to "caffe2_pb2" (in some files both metanet_pb2 and caffe2_pb2 are imported; for these files rename all caffe2_pb2 to metanet_pb2 and then delete the import of metanet_pb2) .

To help me root cause, can you give me ```which protoc``` and ```conda list | grep proto``` and ```python --version```?

**Comment by sclarkson** [comment]
`which protoc`
`/home/sclarkson/anaconda3/bin/protoc`

`conda list | grep proto`
```
libprotobuf               3.4.1                h5b8497f_0  
protobuf                  3.4.1            py36h306e679_0  
```

`python --version`
`Python 3.6.4 :: Anaconda, Inc.`

**Comment by pjh5** [comment]
@sclarkson sorry for the late reply. The default behavior now is use link protobuf statically, and I don't see this problem anymore. Can you update to the latest code (on the Pytorch repo now at https://github.com/pytorch/pytorch.git)? I'll close this issue, but reopen it if you still see this problem.

**Comment by sclarkson** [comment]
Clone from master pytorch repo looks fixed. But, it looks fixed when linking protobuf statically and dynamically. That wasn't what fixed it.

**Comment by pjh5** [comment]
The workaround I mentioned above also fixes this, even when dynamically linking protobuf.

If you're curious, I found that this was caused by python allocating a new object for "NetDef" when it imported metanet.proto (evident by a different memory address of metanet's reference to NetDef than the NetDef referenced by importing NetDef, but I may be misunderstanding python). Since it works with the workaround above, I think it's either related to A) the import statment at the top of metanet.proto might be incorrect Protobuf or B) importing the two protobuf files at different times or in different places is somehow causing the issue. If someone who knows more about protobuf or python wants to take a look then I'd love the help, but for now I'd recommend that people who need to dynamically link protobuf should just copy metanet.proto into caffe2.proto and update all the references.
